AF-2354: Create Initial integration MVP with Work Item Handlers

jesuino opened the pull request:
Merge after https://github.com/kiegroup/appformer/pull/854 is merged

Related PRs:

https://github.com/kiegroup/kogito-tooling/pull/40
https://github.com/kiegroup/appformer/pull/854

tiagobento commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -36,17 +39,40 @@
     private final Promises promises;
     private final WorkItemDefinitionCacheRegistry registry;
+    private final WorkItemDefinitionParser parser;
+    private ResourceContentService resourceContentService;
```
final? 😛

romartin commented:
Adding @inodeman on the review loop, as he'll be working on next kogito - stunner tasks, so better having him updated on all this stuff. thx!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionParser.java:
Please do NOT use `StringBuffer`... at least here. There exist several downsides related to it if you don't use it properly, mostly because it's syncronized. 
So, at least in client side code in general, better moving to `StringBuilder` or even better just using regular java operators on Strings. Even GWT transpilation and the JVM finally will optimizie your StringBuijlder to java operators, mostly, so it makes not much different on resutls. 
But anyway, please avoid using StringBuffer instances...

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionParser.java:
can it be `static` and `private`?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionParser.java:
can it be `static` and `private`?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionParser.java:
can it be `static` and `private`?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionParser.java:
can it be `static` ?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionParser.java:
in general you're also calling `String#trim()` too many times along code below... probably better doing the `trim` once, and passing it accross helper methods to avoid unnecessary computation time

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionParser.java:
can it be `static` and `private`?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -36,17 +39,40 @@
     @Override
     public Promise<Collection<WorkItemDefinition>> call(final Metadata input) {
-        return promises.resolve(registry.items());
+        return promises.create((success, failure) -> {
```
given the service scope and implementation I would say there is no kinda `cache` at all, right? I mean, once editing a process in VSCode, for example, we'll be performing the service queries, to obtain resource data, and also doing the parse operations over it, again and again, even if the work items on the project were not changed, right?
I'm not complaining, that's fine for me now, because as I understand you can add / remove work item definitions at runtime, right? 
Anyway, jsut branistorming and trying to clarify the behavior of these implementations....

jesuino replied:
Hello Roger,

This is something I would have to discuss. Initially there was a cache, but this method seems to be called once, or, at most, twice (sometimes I see it called with an empty content first, and then called with the actual content). Even with cache, users would have to re-open the editor to get the new content from the *.wid file.

I concluded that a cache would not make sense because everything is destroyed/created when the editor is closed/opened.

romartin replied:
ok thanks makes sense to me too, at least for an initial implementation it looks good enought, good to have some brainstorming anyway.. :+1:

jesuino commented:
Hello @romartin ,

Thanks for your valuable feedback. Please see my response inline

> Also I think we should refactor the modules where new WID parser and service are located - Actually I think you put those in the kie-wb-common-bpmn-kogito-runtime module, right? In fact I would like to rename that modlue, probably it's not properly self-explanatory, but the goal for this module is only providing a webapp for the BPMN editor targeting kogito environments (VSCode, Che, GitHub...). So not sure if that makes sense, but I would prefer moving those classes to some other module, probably on the kie-wb-common-stunner-bpmn-client one? 

I moved the parser and its tests to `kie-wb-common-stunner-bpmn-client`. I can also move the service, this is what we need to discuss. I didn't move it because there was a no-op WorkItemDefinitionClientService implementation called WorkItemDefinitionStandaloneClientService - what I did was add a real action to WorkItemDefinitionStandaloneClientService, and I kept the WorkItemDefinitionClientService interface in kie-wb-common-stunner-bpmn-client.

> This way, by default, once using the BPMN editor, it will use your "new" WID approach (parser and service), but on the KIE workbench world, it'll still use the old VFS services...

AFAIK in BC distribution the WorkItemDefinitionClientService implementation is using the BC infrastructure already, hence this should not impact BC at all.

I made the parser improvements that you suggested. Would you please take a look again? See my latest commit -> fb8ae2efd78268ae94719b5cbf0f9ac2c93fed1c

Thanks!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/workitem/WorkItemDefinitionParser.java:
can be `static` too

jesuino replied:
Thank you, added this improvement and also removed an old test that was breaking the build

jesuino commented:
Thanks for your feedback, @romartin 

I built the packages (Chrome and VSCode) and made it available here: https://drive.google.com/file/d/1AHGxzViHx_OdDapdwy8arkKrLyCVzPWk/view?usp=sharing

The VSCode extension I installed and tested, didn't try the Chrome extension from the ZIP above.

Thanks!

jesuino commented:
jenkins retest this

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/test/java/org/kie/workbench/common/stunner/bpmn/client/workitem/WorkItemDefinitionParserTest.java:
```diff
@@ -0,0 +1,167 @@
+        assertEquals(REST_WID_EXTRACTED_PARAMETERS, wid2.getParameters());
+        assertEquals("\"Result\" : new ObjectDataType(),", wid2.getResults());
+
+    }
```
I believe we should assert all 7 definitions

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -36,17 +40,41 @@
+                }
+                return null;
+            });
+        });
```
We miss tests

romartin commented:
Jenkins please retest this

jesuino commented:
@romartin 
Hello Roger,

The tests were failing for this project because we added tests that broke the parser, but after a discussion we decided that we will keep the parser as it is and discuss how it should evolve in a new JIRA. The kogito-tooling PR and this PR are independents. This PR depends on appformer and droolsjbpm-build-bootstrap PRs hence why it is on AF-2353 branch.
 
@jomarko thanks again for your review. I removed the number of WIDs in test and found that category was not being filled, so I fixed this issue and kept only three WIDs in the test, covering all properties.

Let me know if you have any question.

Thanks!

jesuino commented:
Jenkins execute full downstream build

jesuino commented:
Jenkins execute full downstream build

jomarko commented:
jenkins please retest this

jomarko commented:
jenkins execute full downstream build

jesuino commented:
Jenkins execute full downstream build

jesuino commented:
jenkins retest this

jesuino commented:
It was failing due the lack of rebase, which is now done

jesuino commented:
jenkins retest this

jesuino commented:
Jenkins execute full downstream build

romartin commented:
ahh ok @jesuino , thanks for the answers, it makes sense, sorry I got confused.

I'm jsut checking out this work for testing it, I'll let you know something in a while :+1: 

Thx1

romartin commented:
Hey @jesuino 

I jsut found some time to test this stuff, and it still doesn't work for me, either I'm missing something or is there some other issue... lemme explain what I did and plz lemme know: 

I downloaded your [custom vscode extension](https://drive.google.com/file/d/1AHGxzViHx_OdDapdwy8arkKrLyCVzPWk/view?usp=sharing) and installed it. I also created a process which contains a service task (lLog), and added both `bpmn2` and `wid` files in a folder, which I added then in the VSCode workspace.

The process editor is being opened, but the `Task` is not a Log (service task) type, as it is in the source BPMN file. I understand there is not icon for service tasks yet, but it you notice, also when selecting the task, the properties are not the right ones for the `Log` one. No errors found on the console... see:

![Screenshot from 2019-12-11 03-10-47](https://user-images.githubusercontent.com/4602417/70585475-a68cb280-1bc4-11ea-9e0c-e6ac55b72990.png)

PS: Here are my source files - 

Am I missing something? Any ideas?

jesuino commented:
Hello Roger,

Thanks for the feedback.

Sometimes it happens to me for any component and it seems to be a GWT issue
with some sort of cache. For me it shows when I resize the properties pane.
Please let me know if it works for you.

Thanks!



On Tue, Dec 10, 2019 at 11:17 PM Roger Martínez <notifications@github.com>
wrote:

> Hey @jesuino <https://github.com/jesuino>
>
> I jsut found some time to test this stuff, and it still doesn't work for
> me, either I'm missing something or is there some other issue... lemme
> explain what I did and plz lemme know:
>
> I downloaded your custom vscode extension
> <https://drive.google.com/file/d/1AHGxzViHx_OdDapdwy8arkKrLyCVzPWk/view?usp=sharing>
> and installed it. I also created a process which contains a service task
> (lLog), and added both bpmn2 and wid files in a folder, which I added
> then in the VSCode workspace.
>
> The process editor is being opened, but the Task is not a Log (service
> task) type, as it is in the source BPMN file. I understand there is not
> icon for service tasks yet, but it you notice, also when selecting the
> task, the properties are not the right ones for the Log one. No errors
> found on the console... see:
>
> [image: Screenshot from 2019-12-11 03-10-47]
> <https://user-images.githubusercontent.com/4602417/70585475-a68cb280-1bc4-11ea-9e0c-e6ac55b72990.png>
>
> PS: Here are my source files -
>
> Am I missing something? Any ideas?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kiegroup/kie-wb-common/pull/3034?email_source=notifications&email_token=AACX3DF4RFQWGPKZ4BUURULQYBES3A5CNFSM4JQHU4DKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGRT4DI#issuecomment-564346381>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACX3DEZLT3ZZR63RKNOAMTQYBES3ANCNFSM4JQHU4DA>
> .
>


-- 

William Siqueira

Software Engineer

Red Hat

<https://www.redhat.com>

M: +55-11999232248
<https://red.ht/sig>

romartin commented:
Hey @jesuino 

I don't think this is realted to resizing the properties panel... the issue is that the task imported is a regular task, not a Log task type...

I'm closing the properties panel and opening it again and the task remains the same
just a regular task, not a log task.

What am I missing?

romartin commented:
Hey @jesuino

What I suggest is just to use my `bpmn2` and `wid` files attached on the comment above and try to reproduce it.. let's see if I'm missing something or we can find some fix.

BTW - There will be still missing to use the right service task icons. This PR includes the loading for the wid declarations, but there is still missing to use the new Resource API to fetch the icons for service tasks and display them properly in the process editor. Is this something you @jesuino will also help on, in some other ticket//PR too? Otherwise, plz lemme know so we can re-assign it. CC @ederign @porcelli @krisv 

Thx in advance!

ederign commented:
Hi @romartin , regarding your comment 

"BTW - There will be still missing to use the right service task icons. This PR includes the loading for the WID declarations, but there is still missing to use the new Resource API to fetch the right icon for each service task, and display them properly in the process editor. Is this something you @jesuino will also help on, in some other ticket//PR too? Otherwise, plz lemme know so we can re-assign it. CC @ederign @porcelli @krisv"

As we already discussed our implementation is just a draft parser in order to validate the Resource Content API. IMO, any further work or functionality should be prioritised and done by someone on stunner team (@jesuino needs to work on other priorities).

My suggestion is if you are ok with the rest of the code (excluding the parsers), approve the resource content API code, create JIRAs/epic for a 'production ready' version of WID parser and assign to someone of stunner team.

Makes sense?

If it's possible we really would like to merge this tomorrow! Thanks again!

jesuino commented:
Hello @romartin 

Thanks for the clarification. I thought you were facing a caching issue, but then you clarified that the problem was that the task was not recognized.

I made a few improvements in the parser:

* Renamed it to WorkItemDefinitionClientParser to avoid confusion with the server side one;
* Correctly parsed parameters so now the parameters have the correct Java types;
* Added the default category;


This, however, does not fix the issue you are seeing. But I think the problem is not related to the parsing because the generated BPMN with a custom task dragged from the services tasks pallet looks like the same one generated in Business Central. Therefore the issue could be related to Stunner Kogito version because you should be able to reproduce same behavior even if you return a hardcoded list of WIDs in WorkItemDefinitionStandaloneClientService. 

Please let me know if you have any question.

Thanks!

jesuino commented:
Jenkins execute full downstream build

romartin commented:
ok @ederign @jesuino thanks for the updates!

Approving it, although not really sure if we want to merge this yet. I mean, it partially works, it needs a few more love (probably re-implementing the parser and adding support for icons, plus bug fixing), so it's not really yet "ready" IMO for the next kogito release. 

The plan would be next week @inodeman continuing this job, so we have both options: even merging now or keeping this PR open, and then @inodeman can work on top of it and complete it properly during next week. What do you prefer? CC @ederign @porcelli @krisv @lazarotti 

Thanks!

romartin commented:
Hey @jesuino @ederign 

Let you know guys that me and @inodeman are working on [KOGITO-277](https://issues.redhat.com/browse/KOGITO-277), by starting on top of this PR stuff. So it you agree, not still merging because it's still not clear if we'll deliver a new PR or just another one on top yet, does that works / makes sense for you guys?

Thx in advance!

ederign commented:
Sure thing @romartin  feel free to close this or create over the top! Whatever is best for you guys!

romartin commented:
@ederign @jesuino -> here is the continuation of this work https://github.com/kiegroup/kie-wb-common/pull/3087

I'm fine with closing this PR and just having the new one, I'll squash and commit finally, is that ok for you guys?

Thanks in advance!

jesuino commented:
Thanks for the feedback, @romartin 

closing this to focus on #3087
